drivers/mtd_spi_nor/mtd_spi_nor.c

```diff
@@ -0,0 +1,493 @@
+
+const mtd_desc_t mtd_spi_nor_driver = {
+    .init = mtd_spi_nor_init,
+    .read = mtd_spi_nor_read,
+    .write = mtd_spi_nor_write,
+    .erase = mtd_spi_nor_erase,
+    .power = mtd_spi_nor_power,
+};
+
+/**
+ * @internal
+ * @brief Send command opcode followed by address, followed by a read to buffer
+ *
+ * @param[in]  dev    pointer to device descriptor
+ * @param[in]  opcode command opcode
+ * @param[in]  addr   address (big endian)
+ * @param[out] dest   read buffer
+ * @param[in]  count  number of bytes to read after the address has been sent
+ */
+static void mtd_spi_cmd_addr_read(mtd_spi_nor_t *dev, uint8_t opcode,
+    be_uint32_t addr, void* dest, uint32_t count)
+{
+    TRACE("mtd_spi_cmd_addr_read: %p, %02x, (%02x %02x %02x %02x), %p, %" PRIu32 "\n",
+        (void *)dev, (unsigned int)opcode, addr.u8[0], addr.u8[1], addr.u8[2],
+        addr.u8[3], dest, count);
+
+    uint8_t *addr_buf = &addr.u8[4 - dev->addr_width];
+    if (ENABLE_TRACE) {
+        TRACE("mtd_spi_cmd_addr_read: addr:");
+        for (unsigned int i = 0; i < dev->addr_width; ++i)
+        {
+            TRACE(" %02x", addr_buf[i]);
+        }
+        TRACE("\n");
+    }
+    /* Acquire exclusive access to the bus. */
+    spi_acquire(dev->spi, dev->cs, dev->mode, dev->clk);
+
+    do {
+        /* Send opcode followed by address */
+        spi_transfer_byte(dev->spi, dev->cs, true, opcode);
+        spi_transfer_bytes(dev->spi, dev->cs, true, (char *)addr_buf, NULL, dev->addr_width);
+
+        /* Read data */
+        spi_transfer_bytes(dev->spi, dev->cs, false, NULL, dest, count);
+    } while(0);
+
+    /* Release the bus for other threads. */
+    spi_release(dev->spi);
+}
+
+/**
+ * @internal
+ * @brief Send command opcode followed by address, followed by a write from buffer
+ *
+ * @param[in]  dev    pointer to device descriptor
+ * @param[in]  opcode command opcode
+ * @param[in]  addr   address (big endian)
+ * @param[out] src    write buffer
+ * @param[in]  count  number of bytes to write after the opcode has been sent
+ */
+static void mtd_spi_cmd_addr_write(mtd_spi_nor_t *dev, uint8_t opcode,
+    be_uint32_t addr, const void* src, uint32_t count)
+{
+    TRACE("mtd_spi_cmd_addr_write: %p, %02x, (%02x %02x %02x %02x), %p, %" PRIu32 "\n",
+        (void *)dev, (unsigned int)opcode, addr.u8[0], addr.u8[1], addr.u8[2],
+        addr.u8[3], src, count);
+
+    uint8_t *addr_buf = &addr.u8[4 - dev->addr_width];
+    if (ENABLE_TRACE) {
+        TRACE("mtd_spi_cmd_addr_write: addr:");
+        for (unsigned int i = 0; i < dev->addr_width; ++i)
+        {
+            TRACE(" %02x", addr_buf[i]);
+        }
+        TRACE("\n");
+    }
+    /* Acquire exclusive access to the bus. */
+    spi_acquire(dev->spi, dev->cs, dev->mode, dev->clk);
+
+    do {
+        /* Send opcode followed by address */
+        spi_transfer_byte(dev->spi, dev->cs, true, opcode);
+        spi_transfer_bytes(dev->spi, dev->cs, true, (char *)addr_buf, NULL, dev->addr_width);
```

Comment:
This currently keeps the CS pin asserted on erase commands (count = 0), causes unit test failures on Mulle
Fixed:
```c
        bool cont = (count > 0); /* only keep CS asserted when there is data that follows */
        spi_transfer_bytes(dev->spi, dev->cs, cont, (char *)addr_buf, NULL, dev->addr_width);
        /* Write data */
        if (cont) {
```